lib/options.go

```diff
@@ -187,6 +189,103 @@ func ParseCIDR(s string) (*IPNet, error) {
 	return &parsedIPNet, nil
 }
 
+// HostnameTrie is a tree-structured list of hostname matches with support
+// for wildcards exclusively at the start of the pattern. Items may only
+// be inserted and searched. Internationalized hostnames are valid.
+type HostnameTrie struct {
+	children map[rune]*HostnameTrie
+}
+
+// Regex description of hostname pattern to enforce blocks by. Global var
+// to avoid compilation penalty at runtime.
+// Matches against strings composed entirely of letters, numbers, or '.'s
+// with an optional wildcard at the start.
+//nolint:gochecknoglobals
+var legalHostnamePattern *regexp.Regexp = regexp.MustCompile(`^\*?(\pL|[0-9\.])*`)
+
+func legalHostname(s string) error {
+	if len(legalHostnamePattern.FindString(s)) != len(s) {
+		return errors.Errorf("invalid hostname pattern %s", s)
+	}
+	return nil
+}
+
+// UnmarshalJSON forms a HostnameTrie from the provided hostname pattern
+// list.
+func (t *HostnameTrie) UnmarshalJSON(data []byte) error {
+	m := make([]string, 0)
+	if err := json.Unmarshal(data, &m); err != nil {
+		return err
+	}
+	for _, h := range m {
+		if insertErr := t.Insert(h); insertErr != nil {
+			return insertErr
+		}
+	}
+	return nil
+}
+
+// UnmarshalText forms a HostnameTrie from a comma-delimited list
+// of hostname patterns.
+func (t *HostnameTrie) UnmarshalText(b []byte) error {
+	for _, s := range strings.Split(string(b), ",") {
+		if err := t.Insert(s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Insert a hostname pattern into the given HostnameTrie. Returns an error
+// if hostname pattern is illegal.
+func (t *HostnameTrie) Insert(s string) error {
+	s = strings.ToLower(s)
+	if len(s) == 0 {
+		return nil
+	}
+
+	if err := legalHostname(s); err != nil {
+		return err
+	}
+
+	// mask creation of the trie by initializing the root here
+	if t.children == nil {
+		t.children = make(map[rune]*HostnameTrie)
+	}
+
+	rStr := []rune(s) // need to iterate by runes for intl' names
+	last := len(rStr) - 1
+	if c, ok := t.children[rStr[last]]; ok {
+		return c.Insert(string(rStr[:last]))
+	}
+
+	t.children[rStr[last]] = &HostnameTrie{make(map[rune]*HostnameTrie)}
+	return t.children[rStr[last]].Insert(string(rStr[:last]))
+}
+
+// Contains returns whether s matches a pattern in the HostnameTrie
+// along with the matching pattern, if one was found.
+func (t *HostnameTrie) Contains(s string) (matchedPattern string, matchFound bool) {
+	s = strings.ToLower(s)
+	if len(s) == 0 {
+		return s, len(t.children) == 0
+	}
+
+	rStr := []rune(s)
+	last := len(rStr) - 1
+	if c, ok := t.children[rStr[last]]; ok {
+		if match, matched := c.Contains(string(rStr[:last])); matched {
+			return match + string(rStr[last]), true
+		}
+	}
+
+	if _, wild := t.children['*']; wild {
+		return "*", true
```

Comment:
Do we want to allow "*" as a valid pattern? Right now this blocks any requests from being made, which doesn't seem useful or necessary. It also blocks requests to IPs, which seems confusing:

```
WARN[0000] Request Failed                                error="Get \"http://127.0.0.1:8000/\": hostname (127.0.0.1) is in a blocked pattern (*)"
```

I guess it would be useful to block IP ranges, but that seems out of scope for this PR, and it steps on the current `K6_BLACKLIST_IPS` option, so that would need addressing.

So we should probably check whether `host` is an IP in `dialer.go`, and skip hostname-based blocking if it is. This will have a minor conflict with #1489, but it should be easily resolved.

Reply:
> Do we want to allow "*" as a valid pattern? Right now this blocks any requests from being made, which doesn't seem useful or necessary.

I suppose it would be useless - I sought to keep the implementation as literally as possible to what was discussed in #972. I can change this, if deemed necessary.

> It also blocks requests to IPs, which seems confusing ... I guess it would be useful to block IP ranges, but that seems out of scope for this PR, and it steps on the current K6_BLACKLIST_IPS option, so that would need addressing.

I agree. I see a few potential solutions to this problem:
1. Distinguish the dialing process for hostnames and IPs
2. Allow this option to subsume IP blocking
3. Resolve all hostnames to their IPs when reading options, rather than blocking hostnames directly on-dial.

The latter seems more appealing in my opinion, but that would fail to address the [naive blocking method currently employed](https://github.com/loadimpact/k6/blob/master/lib/netext/dialer.go#L84). Another issue I see is that any of these solutions, the two options will end up stepping on top of one another, like you said. Thoughts?

Reply:
I think `3` is not possible really, as hostnames can have multiple IPs and they can change over the course of the test. While currently, k6 caches resolved IPs indefinitely, this is going to change (actually pretty soon if everything works all right ;) ) .

I do think that `1` is the correct way, and I also think that maybe we can leave it for ... after you merge it and we merge #1489 . As far as I can see this will have a slight performance impact, but nothing big. 

The other problem I propose is fixed before the actual trie implementation, but in the config checking - we just don't let the pattern to be .. just `*`. Maybe we even should require it to be at least ~4~ 5 characters as I am pretty sure there are no longer any 1 letter domains ... [actually they are apparently reserved](http://www.nic.ag/rules.htm)